File: runtime/go/units/wasi/fd_pread.go
// requires: memory/i32_load, memory/i32_store, memory/init
func (w *WASI) wasi_fd_pread(fd, iovsPtr, iovsLen uint32, offset uint64, nreadPtr uint32) uint32 {
    f, ok := w.fds[fd].(*os.File)
    if !ok {
        return wasiBadf
    }
    if w.isStdio(f) {
        return wasiSpipe
    }
    nread := uint32(0)
    for i := uint32(0); i < iovsLen; i++ {
        ptr := w.memory.i32_load(uint64(iovsPtr) + uint64(i)*8)
        length := w.memory.i32_load(uint64(iovsPtr)+uint64(i)*8 + 4)
        if length == 0 {
            continue
        }
        buf := make([]byte, length)
        n, err := f.ReadAt(buf, int64(offset+uint64(nread)))
        if n > 0 {
            w.memory.init(uint64(ptr), buf, 0, uint64(n))
            nread += uint32(n)
        }
        // ReadAt returns io.EOF (a non-nil err) on a short read at end of
        // file; either a short read or any error ends the scatter read.
        if err != nil || uint32(n) < length {
            break
        }
    }
    w.memory.i32_store(uint64(nreadPtr), nread)
    return wasiOk
}

File: crates/dewasm-backend-java/units/wasi/fd_pread.java
// requires: memory/i32_load, memory/i32_store, memory/init
// Positional scatter read: reads at an absolute file offset without moving the channel's own position.
// Stdio is not seekable (SPIPE).
int wasi_fd_pread(int fd, int iovsPtr, int iovsLen, long offset, int nreadPtr) {
    Object e = fds.get(fd);
    if (isStdio(e)) {
        return WASI_SPIPE;
    }
    if (!(e instanceof Handle)) {
        return WASI_BADF;
    }
    if (lacksRight(fd, R_FD_READ)) {
        return WASI_NOTCAPABLE;
    }
    java.nio.channels.FileChannel ch = ((Handle) e).ch;
    int nread = 0;
    try {
        for (int i = 0; i < iovsLen; i++) {
            long base = Integer.toUnsignedLong(iovsPtr) + (long) i * 8;
            int ptr = memory.i32_load(base);
            int len = memory.i32_load(base + 4);
            if (len == 0) {
                continue;
            }
            byte[] buf = new byte[len];
            int n = ch.read(java.nio.ByteBuffer.wrap(buf), offset + Integer.toUnsignedLong(nread));
            if (n > 0) {
                memory.init(Integer.toUnsignedLong(ptr), buf, 0, n);
                nread += n;
            }
            // read returns -1 (a short read) at end of file; either ends the scatter.
            if (n < len) {
                break;
            }
        }
    } catch (java.io.IOException ex) {
        return WASI_IO;
    }
    memory.i32_store(Integer.toUnsignedLong(nreadPtr), nread);
    return WASI_OK;
}
